packages/ckb-sdk-core/src/groupInputs.ts

```diff
@@ -0,0 +1,14 @@
+import { scriptToHash } from '@nervosnetwork/ckb-sdk-utils'
+
+type Cell = Pick<CachedCell, 'lock'>
+export const groupInputs = (inputCells: Cell[]) => {
```

Comment:
Quake suggests the name of the group is `ScriptGroup` which be same as `ckb`. 

`ScriptGroup` should only has indexes of every cell input in transaction. When we sign transaction, we can sign each `ScriptGroup` with related private key.

SDK should not know the relationship of `ScriptGroup` and related private keys.

Reply:
Indeed, the pairs of script and private key is specified explicity in the following way:
https://github.com/nervosnetwork/ckb-sdk-js/pull/403/files/719ec3a911ead53cce7198b28a3e43fa4e9c7d1e#diff-458aa15103275f96636c661f9b90d07bR61-R63

The `groupInputs` method only groups the lock scripts so that the first witness of each group could be specified.

Reply:
https://github.com/nervosnetwork/ckb-sdk-js/pull/403/files/719ec3a911ead53cce7198b28a3e43fa4e9c7d1e#diff-458aa15103275f96636c661f9b90d07bR70

We can find that the relationships of inputs and related private key from the example is in sdk core not example

Reply:
https://github.com/nervosnetwork/ckb-sdk-java/blob/develop/example/src/main/java/org/nervos/ckb/MultiSignTransactionExample.java#L124-L143

This example handles the relationships of inputs and private key and `signBuilder.sign(scriptGroup, privateKey)` which is in sdk core.

Reply:
> https://github.com/nervosnetwork/ckb-sdk-java/blob/develop/example/src/main/java/org/nervos/ckb/MultiSignTransactionExample.java#L124-L143
> 
> This example handles the relationships of inputs and private key and `signBuilder.sign(scriptGroup, privateKey)` is in sdk core.

I'm not quite sure but the example is related to `multi sig transaction` while this pr is related to simple `multi inputs transaction`

Reply:
Sorry. Please see this file: https://github.com/nervosnetwork/ckb-sdk-java/blob/develop/example/src/main/java/org/nervos/ckb/MultiKeySingleSigTxExample.java#L147-L169

Reply:
> https://github.com/nervosnetwork/ckb-sdk-js/pull/403/files/719ec3a911ead53cce7198b28a3e43fa4e9c7d1e#diff-458aa15103275f96636c661f9b90d07bR70

```javascript
  const signedTransaction = core.signTransaction(keys)(rawTransaction, cells)
```
The `keys` object includes the relationship between `lockhash` and `privatekey`, which is specified by user in the following format:

```
{
   lockhash1 => privatekey1,
   lockhash2 => privatekey2,
}
```

Reply:
Why lock hash not inputs? lock hash may have many inputs which are not all belong to the transaction.

Reply:
> Why lock hash not inputs? lock hash may have many inputs which are not all belong to the transaction.

Because the inputs are grouped by lockhash.

When the transaction is being signed, SDK will iterate the groups and search the private key by the group's lockhash.

Reply:
I know but I suggest SDK can support a class or interface named `ScriptGroup` which includes the indexes of inputs in transaction and a function whose parameters are scriptGroup and private key. 

BTW. These suggestions are from Quake and I also think it is a more better way.

Reply:
What does `ScriptGroup` mean? Could you provide a reference to the CKB codebase where such struct is?

Reply:
https://github.com/nervosnetwork/ckb/blob/6d48fcad75cd63eff1c3537c19909e6b1ee4fb46/script/src/verify.rs#L43-L51

Reply:
![image](https://user-images.githubusercontent.com/5823268/69600213-ce1f3f00-104a-11ea-803e-2e310e43bbc1.png)

CKB ScriptGroup may be updated later and add other two fileds.

Reply:
> I see but I suggest SDK can support a class or interface named `ScriptGroup` which includes the indexes of inputs in transaction and a function whose parameters are scriptGroup and private key.
> 
> BTW. These suggestions are from Quake and I also think it is a more better way.

Is the point that the method `groupInputs` should be renamed `groupScripts`, which returns the result like:
```javascript
{
    lockhash1 => indices1,
    lockhash2 => indices2,
}
```

Reply:
Yeah. Return the map of key and input indices

Reply:
> Yeah. Return the map of key and input indices

Updated in https://github.com/nervosnetwork/ckb-sdk-js/pull/403/commits/304ba6e3da5169529fe227003e6b9da1c01fdec1